Option to Sort Notes on Export + Config UI

ohare93 opened the pull request:
This PR is an attempt to fix Issue #65 

I may have went a bit overboard from what we talked about, but with this change it is possible to sort by up to 4 keys! :grin: To add another sortable value one need only add a viable option to the sorting_definitions :+1: 


https://github.com/ll-in-anki/CrowdAnki/blob/9834c199ee0e5eb2957d2bf2803bf4ed9e33370b/crowd_anki/export/anki_exporter.py#L101-L109

Let me know what you think!

Still to do
 - Documentation
 - Unit Tests

ohare93 commented:
This PR is an attempt to fix Issue #66 :+1:

Stvad commented on crowd_anki/export/anki_exporter.py:
```diff
@@ -67,3 +71,81 @@ def _copy_media(self, deck, deck_directory):
             except IOError as ioerror:
                 print("Failed to copy a file {}. Full error: {}".format(file_src, ioerror))
+
+    def sort_notes(self, notes):
```
I think this function has grown complex enough to deserve it's own class ;) That'd also simplify testing I believe

ohare93 replied:
Heh, perhaps! Everything else is just an unused MagicMock, so testing is already simple enough. But yes, a separate class would be a lot neater.

Stvad commented on crowd_anki/export/anki_exporter.py:
```diff
@@ -67,3 +71,81 @@ def _copy_media(self, deck, deck_directory):
+
+        # Create mapping of sortable_values variable, to sort_method keys, to sortable_values lambda
+        sorting_definitions = [
+            (0, ("guid", "guids"), lambda i: i.anki_object.guid),
```
I really feel like a separate config validator would better solve the problem you're trying to solve here. Having it inline here adds quite a bit of complexity

ohare93 replied:
True true, I'll get on that.

Stvad commented on crowd_anki/export/anki_exporter.py:
```diff
@@ -67,3 +71,81 @@ def _copy_media(self, deck, deck_directory):
+            if not is_reversed:
+                pass
+            else:
+                notes = list(reversed(notes))
```
I believe we should not do anything if the sort set to `none`.

Stvad commented on crowd_anki/export/anki_exporter.py:
```diff
@@ -67,3 +71,81 @@ def _copy_media(self, deck, deck_directory):
+            if all_methods_good:
+                sort_numbers = [sort_map[sm] for sm in sort_method]
+
+                if len(sort_numbers) == 1:
```
why can't this be a loop/comprehension?

ohare93 replied:
I wanted it to be so, but could not see how to create a tuple of various length based on the length of sort_numbers. Is that possible?

Stvad replied:
Use tuple comprehension? Or create a list and convert it to tuple

Stvad commented on crowd_anki/export/anki_exporter.py:
```diff
@@ -67,3 +71,81 @@ def _copy_media(self, deck, deck_directory):
+            note.export_filter_set.add("sortable_values")
+
+            for defs in sorting_definitions:
+                note.sortable_values[defs[0]] = defs[2](note)
```
if I understand this correctly you apply all key functions in advance here? why do it here and not inline?

also, please refrain from modifying the notes, if you really need to inject additional information for sorting - look into creating a list of pairs, where you sortable information would be in one of the parts of the pair, and note would be the other. Then you'd sort that, extract notes and replace the notes list in the deck (but really I think you should just apply keys inline)

ohare93 replied:
I saw that you had lots of other ignored values on the notes and decided to go with your convention. A paired list is a good idea. Where exactly do you mean when you say "inline"? As in, when the notes are created? If so, the simple answer is that since this code was all kept within this function I had no way (or intention) of making the values available to other parts of the code. Now that I am going to separate it into it's own class that should be doable though. I'll make a sortconfig file which will contain the definitions, which can be accessed in other parts if need be.

Stvad replied:
From the phone on the plane now, but will try to answer in the time I have :p. 
Another way to put this is why not create a lamda that would apply multiple functions to value and return a tuple, instead of just returning values precomputed here?

Stvad replied:
so to give you an example of what I'm talking about, I'd create a sort key function like:

```
# assuming a simple name -> key map

keys = (sorting_definitions[method_name] for method_name in sort_method)
def sort_key(note):   
    return (key(note) for key in keys)

notes = sorted(notes, key=sort_key...)
```

Stvad replied:
>I saw that you had lots of other ignored values on the notes and decided to go with your convention. 

I'm not really fond of those, they are mainly there as a way to filter out the object variables, but I don't think I inject the new values dynamically anywhere..

ohare93 replied:
This was very helpful, thank you. Returning the value inline is a good idea, rather than my ham-fisted approach!

I was trying to do Tuple comprehension before to accomplish this, just like how you have given as an example, but it would never work like I expected. Your suggestion here got me to try again, and after some digging I found that `return (key(note) for key in keys)` will not return a Tuple of keys, but instead a Generator expression. You need to be explicit instead with `return tuple(key(note) for key in keys)`. So my google-fu found :sweat_smile: https://stackoverflow.com/questions/16940293/why-is-there-no-tuple-comprehension-in-python

Anyways I have pushed a change, I hope the result is nicer to your Python sensibilities :smirk: (it is to mine!)

Stvad commented:
> A config validator would be acceptable 😁

It can even do some guessing if you'd like (I'm not convinced that it's really required besides basic .strip.tolower), as long as it's not inlined with the logic of sorting.

A somewhat relevant link on the topic: https://en.wikipedia.org/wiki/Levenshtein_distance

ohare93 commented:
Pushed a change, all the functionality is now in a NoteSorter class. I think you'll agree it's a lot nicer now, all the settings are calculated in the constructor, and the `sort_notes` function is down to 16 lines long!

Still to do:
 1) Move config settings out of NoteSorter :thinking: 
 1) Unit Tests for NoteSorter(much easier now!)
 1) Documentation

May do for this PR:
 - Make a config validator, which occurs on change of the config settings

I have some other changes I wish to implement to CrowdAnki, such as fixes for Issue #31 and Issue #23. Those would need to use the config settings to some degree, and thus would benefit from an improved custom UI for the config, which would also easily allow for validation. I too would like to remove the validation from the NoteSorter, but I do not know if I will get around to it for this PR if the config UI takes too long. If not, I will remove it as part of one of these other changes in the coming month :+1:

ohare93 commented:
Got some help from a friend who's also working on Anki extensions. He steered me in the right direction for QT UI files, and so it was simple to make one for the config page. Behold (a work in progress)!

![image](https://user-images.githubusercontent.com/1471350/67165595-51cc7880-f387-11e9-92a3-32124a15e649.png)

I'm still connecting up the text boxes to their respective config values, so the page is still a work in progress. But the checkboxes are fully functional, and change their config values.

When the user hits "Okay" then validation can be done, and will not close the window or save the info the config if the values are incorrect. I have made dedicated classes and enums for the Config data info, and Sort Method options, which will make that simple.

Let me know what you think :grin:

ohare93 commented:
Update: all config values are now connected up through the GUI to the config file itself :+1: 

![image](https://user-images.githubusercontent.com/1471350/67186199-db139780-f3e7-11e9-9946-a97aaa493d8d.png)

The Note Sort Type box also has input verification upon hitting "Ok", not allowing the user to save until only correct values remain.

![image](https://user-images.githubusercontent.com/1471350/67186298-1a41e880-f3e8-11e9-82da-a0374007ca7d.png)

It will be a lot easier in the future to add config values :grin: Simply add a checkbox/textbox in QT Designer, build the UI file to Python, connect the ui items to the config file itself in `config_settings.py`, then just use it!

ohare93 commented:
Spread the ConfigSettings class to the rest of the code, now the ActionVendor and HookVendor simply use the values ConfigSettings class which is passed to them, which handles all the default values and reading the config file and whatnot.

Also finished some Unit Test for each filter options, to make sure it returns the cards in the right order :+1: 

I believe with this I am 99% done! :grin: I'm sure there's some more bits and bobs lying around, but please do give some feedback (whenever you're not too busy with travelling) :+1:

Stvad commented:
wow, that's a lot of nice things :) I'm heading back home now, so should be able to give you some feedback soon.

Stvad commented on CONTRIBUTING.md:
```diff
@@ -11,6 +11,12 @@ of Anki.
+The config is build using QTDesigner, which you can get by running `sudo apt install qtcreator` on Linux, or simply download from `https://build-system.fman.io/qt-designer-download`
+
+In QT Designer you can create UI files using a simple GUI, then use the `pyuic5` module to convert them to python:
+`pyuic5 ui_files/config.ui -o crowd_anki/config/config_ui.py`
```
this should be part of build/packaging process (see shell script in the root)

Stvad replied:
https://github.com/Stvad/CrowdAnki/blob/master/package_plugin.sh

ohare93 replied:
Oh yes, that's a good idea, I'll add it there.

In the future if more UI files are added then switching over to using [Glutanimate's AnkiAddonBuilder](https://github.com/glutanimate/anki-addon-builder) should be considered. It builds all the UI files in the repo for you, and packages the add-on for both(/either) Anki 2 and 2.1. I have no idea if there is some dependency which means this project cannot use that, and I am also not wedded to the idea, it's just a thought :thinking:

Stvad replied:
also pyuic5 should be added to build dependencies

Stvad replied:
I believe this one needs to be updated to reflect changes in automation around it :)

ohare93 replied:
Oh yes, I updated the documentation too.

Stvad replied:
👍 It bothers me that people may need to copypaste things from here though and more details then I like are exposed 🤔.

How about moving this line invocation into a separate script like `generate_ui.sh` that would install the pipenv dev deps (to get the pyuic5) and runs this. Or maybe just runs this, depending on whether repeatedly running pipenv takes long time.

And mentioning it here + using it from the other scipt?

+ should probably mention installing pipenv dev deps as a way to get the `pyuic5` in general

Also it's `QtDesigner` with small `t` 😛

ohare93 replied:
Good call, changed :+1: added `generate_ui.sh`, called it from the packaging script, and rewrote the documentation :+1:

Stvad commented on crowd_anki/anki/hook_vendor.py:
```diff
@@ -4,17 +4,13 @@
 @dataclass
 class HookVendor:
     window: Any
+    config: object
```
should be of a type of the class you've created (here and elsewhere)

ohare93 replied:
I left it as an object in the places which don't actually use it, they only pass it on. But I can change that if you prefer.

Stvad replied:
as we need to specify some type, we might as well specify a correct one here ;)

ohare93 replied:
Done

Stvad commented on crowd_anki/config.json:
```diff
@@ -1,4 +1,3 @@
 {
-  "automated_snapshot": false,
```
why?

ohare93 replied:
This file is not needed to set a user default anymore. The default values are set on each property in the config class. Adding further properties there would only add a requirement to add them here (and adding them here would still do nothing, as nothing reads this).

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+    export_deck_sort_methods: list
+
+
+    def __init__(self, get_settings = True):
```
why do we have this option here? when would we want not to load them?

ohare93 replied:
That is for the unit tests, didn't want them to load the values but for the tests to set them manually.

Stvad replied:
so this'll depend on what we'll eventually do with the `config_file` but I think a better option here may be to accept initializer for that dict and if it's none load it from anki, that'd imho proved and more natural/full way to test the thing

ohare93 replied:
This can be an improvement for the future, sure.

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+
+    def __init__(self, get_settings = True):
+        if get_settings:
+            self.config_file = mw.addonManager.getConfig(__name__)
```
I believe this is problematic. This means that we cache the settings on the creation of the object (which would happen on plugin startup, at least in some cases), so to get new values a person would have to restart Anki, which is not a great experience. 

I think the config_file should be a cached property with a small ttl (https://pypi.org/project/cached-property/) or even just a function that gets the file each time

ohare93 replied:
The ConfigSettings class is the only way to read the config values, yes. But it is also the only way to set them yourself now. Opening the config options directly interacts with this class, and everywhere in the code relies on this class to give the values. It is made in the main and passed everywhere else, so it is pretty much a global singleton. If I have missed something in this implementation then I can go back and rethink this, but I do not currently see a time where a users value could be cached so that their intended value as not set. Unless they change the settings json with some other file and not through Anki? But 1. Do we wish to support that? and 2. the settings in the config window will not reflect their changes so it will be easy to see nothing has changed.

Stvad replied:
The thing is - no one prevents people from creating multiple instances of the class. 's not actually a singletone (though making it such can be a solution).

A simple example of how it breaks: 
* Launch anki, go to config, enable automated snapshot. 
* What I expect to happen: when I quit anki/switch profile automated snapshot happens
* What actually happens: automated snapshot does not happen until the Anki is restarted and the new config values are loaded.

The previous config usages that I had were broken in a similar fashion on reflection. 

Some solutions for this would be to:
* do the cached property thing I mentioned above 
* make the config class an actual singleton

Stvad replied:
what is you current thinking on this?

Stvad commented on crowd_anki/main.py:
```diff
@@ -7,10 +7,28 @@
+    """
+    Launch custom GUI on config change instead of default Anki JSON editor
+    """
+    mw.crowd_anki_config = config = ConfigDialog()
```
why do we need to inject a variable on the mw for it?

ohare93 replied:
It is suggested to do this in the [Writing Anki 2.1 Add-ons documentation](https://apps.ankiweb.net/docs/addons.html#qt), so that the garbage collection does not delete the window when it is not used.

Stvad replied:
I see.

Stvad replied:
in the current implementation though we'd recreate the widget from scratch each time anyhow (as we call a constructor here each time config action is invoked). So doing this does not really give us anything

ohare93 replied:
Changed to singleton :+1:

Stvad replied:
err, this is a different discussion I believe ;)

ohare93 replied:
Ahh I understand your point now! Okay I have moved the line to the initialize_config_window function, and just left the invoke_config_window to execute it. Like this:

```

def invoke_config_window():
    """
    Launch custom GUI on config change instead of default Anki JSON editor
    """
    mw.crowd_anki_config.exec_()


def initialize_config_window():
    """
    Add option for addon's config in Anki
    :return:
    """
    mw.crowd_anki_config = ConfigDialog()
    mw.addonManager.setConfigAction(__name__, invoke_config_window)

```

 Tested and it works :+1: Does this sound better? :grin:

Stvad replied:
this one is fine :) , creating a new widget each time was probably also fine tbh - it just didn't make sense to me to both create it each time and store the reference as advised by the link you give 😛

Stvad commented on crowd_anki/meta.json:
```diff
@@ -0,0 +1 @@
+{"config": {"automated_snapshot": false, "snapshot_root_decks": [], "export_deck_sort_reversed": false, "export_deck_sort_method": ["none"], "snapshot_path": "/home/jmo/Development/CrowdAnki/crowd_anki/user_files"}}
```
prettify this one? (e.g. newlines, indentations/etc)

also, you've commited your snapshot path

Stvad replied:
also, can't comment there, but `zzzNoteTypeExamples` probably wasn't meant to be commited either?

ohare93 replied:
Yes and yes! :sweat_smile:

ohare93 replied:
Alright, deleted my personal files. But the meta.json is just the local copy of the settings, it gets made everytime the config saves. As it's not in your own master branch and there is no use in having it in source control I deleted it and added it to the gitignore :+1:

ohare93 commented:
> wow, that's a lot of nice things :) I'm heading back home now, so should be able to give you some feedback soon.

Ayy you're back! Hope the holiday went well :+1:

ohare93 commented:
Made the changes you suggested :+1: I expect you aren't properly home yet and haven't played around with the new stuff, so give it a shot whenever you're free and get back to me. When I am assured this PR will/does go through and the new Config settings will be used in their current form then this new easier way of creating config settings will allow me to quickly and easily make a few others changes. 

I plan to:
1) Make a simple fix for #23 with just a checkbox in the config
2) Try and tackle my own issue #62 -- no idea how long this one may take, or what the config options will look like until I try it
3) See what I can do about issue #31. This one I imagine you will have lots of opinions about, as it's simple enough to give the user the option to not export deck headers, but then I'd need to make it possible to import without them, and also perhaps add in a config setting for "Ignore deck header on import". Lots of possible things here.

Issue #22 looks simple enough to fix too. Anyways, thanks for all the feedback, and for making this useful program in the first place! I hope I can do whatever I can to make it a bit better :grin:

Stvad commented:
Just loaded the UI, some notes: 
* neat!
* how about using a multi-select for sort options (e.g. with List widgit or something). (The root decks should probably be the same, but that's likely more involved and out of scope here)
* how about using QFileDialog for selecting the snapshot path?
* section header labels should probably be larger, they are dominated by other elements on my screen

<img width="1748" alt="image" src="https://user-images.githubusercontent.com/81347/67732150-b014df00-f9b7-11e9-9938-0344ae024327.png">

Stvad commented:
btw, as I loaded it locally - Pycharm tells me that there is a variety of pep8 violations - please look into fixing those

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+from ..utils.constants import USER_FILES_PATH
+
+@dataclass
+class ConfigSettings():
```
nit: the braces here are redundant

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+        FLAG = "flag"
+        TAG = "tag"
+        NOTEMODEL = "notemodel"
+        NOTEMODELID = "notemodelid"
```
add `_` to separate words e.g. `NOTEMODEL` -> `NOTE_MODEL`

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+        FIELD1 = "field1"
+        FIELD2 = "field2"
+
+    snapshot_path: str
```
I think the fields should go first in the class

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+
+    def __init__(self, get_settings = True):
+        if get_settings:
+            self.config_file = mw.addonManager.getConfig(__name__)
```
btw, 's not actually a file, 's a dictionary, so the current name is confusing. `config_dict` maybe? or just `_config`

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+            self.get_all_values()
+
+    def get_all_values(self):
+        self.snapshot_path = self.config_file.get(self.Properties.SNAPSHOT_PATH.value, str(USER_FILES_PATH.resolve()))
```
I think it'd better if the default values would be explicitly defined in the property enum. So do something like:

```
ConfigEntry = namedtuple("ConfigEntry", ["name", "default_value"])

class Properties(Enum):
        SNAPSHOT_PATH = ConfigEntry("snapshot_path", str(USER_FILES_PATH.resolve()))
...
```

Stvad replied:
Also this would benefit from a bit of metaprogramming. Something like:

```
    def _get(self, prop: Properties):
        self.config_file.get(prop.value.config_name, prop.value.default_value)

    def load_values(self):
        for prop in self.Properties:
            setattr(self, prop.value, self._get(prop))
```

Stvad replied:
hmm, you don't need to use getattr to access them in this case. As a minimal scenario something like this works:

```python
class Dummy:
    pass

x = Dummy()

setattr(x, "name", "value")

print(x.name)
```

with type hints you can make it more obvious what values your class can vend + keep autocompletion working, so you can do something like this:

```python
@dataclass
class Dummy2:
    name: str = field(init=False)

y = Dummy2()

setattr(y, "name", "value2")

print(y.name)
```

Stvad replied:
Ah I see, I believe it wasn't pep8 though, but Pycharm complaining that it can't figure out what those values are (which as you mentioned is fair 😛). The approach 2 above should remediate that though.

ohare93 replied:
Is there any benefit to changing the current class to a dataclass with nothing initialised? There are already no warnings :wink: 

```
class ConfigSettings:
    snapshot_path: str
    automated_snapshot: bool
    snapshot_root_decks: list
    export_notes_reverse_order: bool
    export_note_sort_methods: list

```

Stvad replied:
Nope, given that we don't use the dataclass functionality it's actually better like it is right now - just with type hints :)

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+
+
+    def set_all_values(self):
+        self.config_file.update({
```
this would also benefit from some metaprogramming similar to above but with getattr

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+            self.Properties.EXPORT_DECK_SORT_METHODS.value: self.export_deck_sort_methods
+        })
+
+    def write_values_to_anki(self):
```
should this be called something like `save`?

Stvad replied:
I thought about just `save` really 😛

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,75 @@
+            self.error_message_popup.setText(f"Invalid config values: \"{message_text}\"")
+            self.error_message_popup.open()
+        else:
+            self.config.set_all_values()
```
this seems like an internal detail that should be called from `write_values_to_anki`

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+    def find_invalid_config_values(self):
+        self.handle_empty_textboxes()
+        
+        invalid_methods = [method for method in self.export_deck_sort_methods if method not in self.DeckExportSortMethods._value2member_map_]
```
this should be split in 2 lines.

Also is the `_value2member_map_` guaranteed by enum api or is it an implementation detail?

Stvad replied:
also you can probably inline the `invalid_methods`

Stvad replied:
hmm, I don't see it in doc https://docs.python.org/3/library/enum.html can you give me a link?

ohare93 replied:
:thinking: no I seemingly cannot. Hmm.

Stvad replied:
So seems more like an implementation detail. I'd prefer we stay away from using those in favor of public API's

Stvad replied:
you can still basically do what you did before, just obtain the values in a different way

for example you can add the following method to the `NoteSortingMethods` enum:

```
    @classmethod
    def values(cls):
        return set(it.value for it in cls())
```

and then use `NoteSortingMethods.values()` where you've used the map before

ohare93 replied:
Ah good call. Changed back :+1: :smirk:

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+            self.export_deck_sort_methods = [self.DeckExportSortMethods.NO_SORTING.value]
+        
+        if not self.snapshot_path:
+            self.snapshot_path = str(USER_FILES_PATH.resolve())
```
how about making these properties instead and moving this logic to corresponding getters?

Stvad replied:
hmm, not entirely sure it'd solve this, it'd provide the default values sure, but presumably you still want the check logic?

ohare93 replied:
I'll make your other change soon and we'll see.

ohare93 replied:
Fixed, as suggested. I just made a "set to default" function which takes a property. I only want it to be like this for these two text boxes, and I don't want it to be on all of them by default. Let me know and tell me what you think.

Stvad replied:
's not quite what I meant. I had in mind having getters for this specific values that would encapsulate logic for retrieving their defaults. Something like:

```python
@attribute
def snapshot_path():
    return self._get(self.Properties.SNAPSHOT_PATH)

# so now if someone calls 

config.snapshot_path

# this function would be called
```

ohare93 replied:
Ah I see. Well, while trying to do this for each I encountered an issue: I need to set these values *locally* as a config_settings property, not in the proper config set. They only go in the config when they are being accepted and saved. For example when someone enters "guid, testindfsdkfjsd" as a sort method, this is saved as the local value of export_note_sort_methods in the config, so that it can perform verification on it on save. So since the UI files need to set them, I would have to also make a @property.setter on each, but they all are do damn simple apart from the export_note_sort_methods. 

So I settled on simply adding back in the original names properties, which will be set by load_values anyways, and adding a formatted_export_note_sort_methods property, which is called in NoteSorter now.

```python
class ConfigSettings:
    snapshot_path: str
    automated_snapshot: bool
    snapshot_root_decks: list
    export_notes_reverse_order: bool
    export_note_sort_methods: list

    @property
    def formatted_export_note_sort_methods(self) -> list:
        return [
            NoteSortingMethods(method)
            for method in self.export_note_sort_methods
        ]
```

Thoughts on this?

Stvad replied:
Sure, some thoughts:

* You don't have to do it for all the fields, only for the ones that require custom verification logic/etc, and you can leave the other ones as is
* Also setters/getters can as well be automatically generated similar to how the fields are now. 
* The values are not written to file until you actually call `writeConfig` so I'm not sure that you actually need the intermediate storage

Stvad commented on crowd_anki/export/anki_exporter.py:
```diff
@@ -5,29 +5,37 @@
 
 class AnkiJsonExporter(DeckExporter):
-    def __init__(self, collection,
+    def __init__(self, collection, config: ConfigSettings,
```
nit: newline

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+    sort_key_tuple: tuple
+
+    @classmethod
+    def from_config(cls, config: ConfigSettings):
```
should this just be the __init__ method?

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+    @classmethod
+    def from_config(cls, config: ConfigSettings):
+        sorting_definitions = {
+            ConfigSettings.DeckExportSortMethods.NO_SORTING: None,
```
this looks combersome, we're probably better off extracting the enum from the configsettings class to reduce the noise here.

Stvad replied:
I meant extracting the DeckExportSortMethods to be a top-level enum and not be part of ConfigSettings

ohare93 replied:
Done.

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+        EXPORT_DECK_SORT_REVERSED = "export_deck_sort_reversed"
+
+
+    class DeckExportSortMethods(Enum):
```
`NoteSortingMethods` maybe?

ohare93 replied:
Same as above.

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+        }
+        
+        cls.sort_method = [ConfigSettings.DeckExportSortMethods._value2member_map_[method] for method in config.export_deck_sort_methods]
+        cls.is_reversed = config.export_deck_sort_reversed
```
unless I'm missing something I don't believe you want to assign anything to the class instance

Stvad replied:
sure, you can, but why would you want to in that case?

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+            ConfigSettings.DeckExportSortMethods.FIELD2: lambda i: i.anki_object.fields[1]
+        }
+        
+        cls.sort_method = [ConfigSettings.DeckExportSortMethods._value2member_map_[method] for method in config.export_deck_sort_methods]
```
I think the `config.export_deck_sort_methods` should return enum values straight away. Also the conversion to enum from value can be done by doing `DeckExportSortMethods(value)` afaik.   
Also I think it may make sense to make name and value the same for `DeckExportSortMethods`

Stvad replied:
or you can make them case-insensetive and do `.upper` on all the strings

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+
+@dataclass
+class NoteSorter():
+    sort_method: list
```
`sort_methods`

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+    sort_method: list
+    is_reversed: bool
+    skip_sorting: bool
+    sort_key_tuple: tuple
```
not sure if you need this separately from `sort_methods`

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+        cls.sort_method = [ConfigSettings.DeckExportSortMethods._value2member_map_[method] for method in config.export_deck_sort_methods]
+        cls.is_reversed = config.export_deck_sort_reversed
+
+        cls.skip_sorting = True if cls.sort_method[0] == ConfigSettings.DeckExportSortMethods.NO_SORTING and not cls.is_reversed else False
```
should probably just a method inside the class

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+            print([note.anki_object.guid for note in notes])
+
+        if not self.skip_sorting:
+            if self.sort_method[0] == ConfigSettings.DeckExportSortMethods.NO_SORTING:      # Only the first method is considered for these pass variables
```
why this special case?

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+
+        if not self.skip_sorting:
+            if self.sort_method[0] == ConfigSettings.DeckExportSortMethods.NO_SORTING:      # Only the first method is considered for these pass variables
+                if not self.is_reversed:
```
I think it'd be cleaner to do sort, then reverse

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,60 @@
+                notes = sorted(notes, key=self.get_sort_key, reverse=self.is_reversed)
+
+        if DEBUG:
+            print([note.anki_object.guid for note in notes])
```
should these be done via logging module?

Stvad replied:
https://docs.python.org/3/library/logging.html

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+        AUTOMATED_SNAPSHOT = "automated_snapshot"
+        SNAPSHOT_ROOT_DECKS = "snapshot_root_decks"
+        EXPORT_DECK_SORT_METHODS = "export_deck_sort_method"
+        EXPORT_DECK_SORT_REVERSED = "export_deck_sort_reversed"
```
probably should have the "note" in the name instead of "deck". Something like `EXPORT_NOTE_SORT_METHODS` and `EXPORT_NOTES_REVERSE_ORDER`

ohare93 replied:
I begrudgingly agree :smirk: Will change this soon

Stvad commented on package_plugin.sh:
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env bash
 
+pyuic5 ui_files/config.ui -o crowd_anki/config/config_ui.py
```
as mentioned above the pyuic5 should be part of dev dependencies, and the installation of them locally should probably be done in `fetch_dependencies`

ohare93 replied:
Alright dev it is. Think I have the right one now, after looking into it.

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,75 @@
+        self.ui_initial_setup()
+
+    def accept(self):
+        invalid_values = self.config.find_invalid_config_values()
```
this one is more of an opinion, but something like `self.config.validate()` that throws an exception with problem description + try block here that'd display that error if it's thrown would probably work better here

ohare93 replied:
Definitely opinion, but well received :grin: I try and avoid try catch blocks when I can. This method works fine for now, and I expect when there are other values in the config then we may wish to run some other arbitrary code without interrupting the flow. If you don't mind I'll leave it the way it is for now.

Stvad replied:
well presumably you won't be interrupting the flow, as you'd wrap the validate in the try 😛.

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,75 @@
+    def accept(self):
+        invalid_values = self.config.find_invalid_config_values()
+        if invalid_values:
+            print(invalid_values)
```
this one is probably redundant, given that we display it

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,75 @@
+            super().accept()
+
+    def reject(self):
+        print("Reject")
```
still needed?

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,75 @@
+        self.form.cb_reverse_sort.setChecked(self.config.export_deck_sort_reversed)
+        self.form.cb_reverse_sort.stateChanged.connect(self.toggle_reverse_sort)
+    
+        self.form.textedit_deck_sort_methods.appendPlainText(self.get_formatted_comma_separated_string(self.config.export_deck_sort_methods))
```
should be split in multiple lines. also why append* and not set* (not sure what the appropriate method would be)

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,75 @@
+        self.config = ConfigSettings()
+        self.form = ConfigUI()
+        self.form.setupUi(self)
+        self.ui_initial_setup()
```
`setup_from_config`?

ohare93 replied:
Are you suggesting to change the `ui_initial_setup` function to `setup_from_config`? The current name makes more sense to me: it sets up the initial values and onchange providers of all the UI parts.

Stvad commented on test/anki/hook_vendor_spec.py:
```diff
@@ -6,14 +6,26 @@
+        hook_manager = MagicMock()
+        config = ConfigSettings(False)
+        config.automated_snapshot = True
+        vendor = HookVendor(window=MagicMock(), config=config, hook_manager=hook_manager)
```
extract common setup between these tests to a function

ohare93 replied:
Sure, done :+1:

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+        return note
+
+    def setup_notes(self):
+        random_range = list(range(0, 6))
```
tie this to size of data arrays/constant

Stvad replied:
also with random things it's a good idea to print the current configuration you're working with (the shuffled number list in this case) to make any failures easier to reproduce

ohare93 replied:
Sure, but then it's just tied to one of them. I chose `test_guids`

ohare93 replied:
> also with random things it's a good idea to log the current configuration you're working with (the shuffled number list in this case) to make any failures easier to reproduce

Good call.

ohare93 replied:
As I asked in another comment, which level of logging would you do here?

Stvad replied:
info is fine

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+    def get_single_note_mock(self, i):
+        note = MagicMock()
+
+        note.anki_object.guid = test_guids[i]
```
assign `note.anki_object` to a variable? or initialize it via dict maybe

ohare93 replied:
I thought the MagicMock took care of that. It doesn't seem to mind :sweat_smile:

Stvad replied:
sure it does not mind, I though it may be cleaner that way though ;)

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+
+        note.anki_object._model = {}
+        note.anki_object._model.setdefault("name", test_notemodels[i])
+        note.anki_object._model.setdefault("crowdanki_uuid", test_notemodelids[i])
```
initialize it as a dict

ohare93 replied:
Oh yea, doh. Done.

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+
+        return notes_list
+    
+    def set_sort_method(self, set_to):
```
i find the name of parameter confusing

ohare93 replied:
:thinking: it sets the sort_method :sweat_smile: what would you change it to?

Stvad replied:
parameter not a function. I'd do something like `sort_method`

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+from crowd_anki.config.config_settings import ConfigSettings
+
+
+test_guids = ["aaabbbccc", "bbbcccddd", "cccdddeee", "dddeeefff", "eeefffggg", "fffggghhh"]
```
"abc", "bcd" etc would probably be preferable, as the current list makes me start thinking on whether they are the same length, etc, and it's a bit harder to visually compare

ohare93 replied:
Sure :+1:

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+    def set_sort_method(self, set_to):
+        self.config.export_deck_sort_methods = set_to
+    
+    def setup_note_sorter(self):
```
given the class name - just `setup` probably

ohare93 replied:
Yea sure :+1:

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+        self.sorted_notes = self.notesorter.sort_notes(self.notes)
+    
+
+    def assert_sorted_notes(self, equals):
```
can probably inline this one

ohare93 replied:
Had plans to add to it, but yea sure. Inlined :+1:

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+        with it("sort by guid"):
+            self.tester.set_sort_method([ConfigSettings.DeckExportSortMethods.GUID.value])
+            
+            self.tester.setup_note_sorter()
```
given that you always call this one, it should probably be part of the `set_sort_method`

ohare93 replied:
I'd rather not, that kind of mixes up the intentions of each. Also I expect there'll be different types of tests here needed later, which may need to setup multiple things. It's clear what it does right now, I'm going to leave it as it is.

Stvad replied:
hmm, maybe make `setup` accept the sort methods, or do a third method that'd call both (though setup seems to be pretty well situated for this)?

ohare93 replied:
Done.

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+            
+            self.tester.setup_note_sorter()
+
+            assert([note.anki_object.guid for note in self.tester.sorted_notes] == test_guids)
```
you can probably reuse the key from map elsewhere fro this query, have a map from type to expected value and then run over all the cases in 1 loop.

ohare93 replied:
I don't understand your suggestion. Precalculate the sorted_notes values? Well do remember the unit test randomises the order of the values each time.

Stvad replied:
No, I meant doing something like:
```
method_result_pairs = [(DeckExportSortMethods.FLAGS, test_flags),(DeckExportSortMethods.GUID, test_guids)]

for method, result in method_result_pairs:
  [NoteSorter.sorting_definitions[method](note) for note in  self.tester.sorted_notes] == result
```

Stvad replied:
the logging module can be configured w.r.t what to print and not to print, also you can specify file as destination for logs

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,122 @@
+test_fields = test_tags
+
+
+class NoteSorterTester():
```
you're not testing multi-key key case

ohare93 replied:
Yea need to do that.

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,74 @@
+from ..utils.constants import USER_FILES_PATH
+
+@dataclass
+class ConfigSettings():
```
some tests for this one would be nice

ohare93 replied:
It would. I may get around to that.

Stvad commented:
> Made the changes you suggested 👍 I expect you aren't properly home yet and haven't played around with the new stuff, so give it a shot whenever you're free and get back to me. When I am assured this PR will/does go through and the new Config settings will be used in their current form then this new easier way of creating config settings will allow me to quickly and easily make a few others changes.
> 
> I plan to:
> 
>     1. Make a simple fix for #23 with just a checkbox in the config
> 
>     2. Try and tackle my own issue #62 -- no idea how long this one may take, or what the config options will look like until I try it
> 
>     3. See what I can do about issue #31. This one I imagine you will have lots of opinions about, as it's simple enough to give the user the option to not export deck headers, but then I'd need to make it possible to import without them, and also perhaps add in a config setting for "Ignore deck header on import". Lots of possible things here.
> 
> 
> Issue #22 looks simple enough to fix too. Anyways, thanks for all the feedback, and for making this useful program in the first place! I hope I can do whatever I can to make it a bit better 😁

awesome! I really appreciate your contributions to making CrowdAnki better! 
One thing for the future - I'd advise splitting up the work in a smaller chunks so we both can iterate faster on it 🙂

ohare93 commented:
I somehow missed this comment! 😅 

> * how about using a multi-select for sort options (e.g. with List widgit or something). (The root decks should probably be the same, but that's likely more involved and out of scope here)

Yea that's a lot harder than it sounds, since you can have a variable number of sort options. Either multiple sort options multi-select boxes are shown (an explicit set amount) or new ones are generated as methods are selected 😓  (and checks are made to not be able to select the same thing in each)

> * how about using QFileDialog for selecting the snapshot path?

Good idea 👍 

> * section header labels should probably be larger, they are dominated by other elements on my screen
> 
> <img alt="image" width="1748" src="https://user-images.githubusercontent.com/81347/67732150-b014df00-f9b7-11e9-9938-0344ae024327.png">

Jesus that is not how it looks on my screen! 😅 Here's how it looks to me:

> ![image](https://user-images.githubusercontent.com/1471350/67186199-db139780-f3e7-11e9-9946-a97aaa493d8d.png)

Making the headers bigger should be fine though.

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,39 @@
+from ..config.config_settings import ConfigSettings
+
+
+class NoteSorter():
```
nit redundant parenthesis

ohare93 replied:
Removed

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,76 @@
+    def changed_textedit_snapshot_path(self):
+        self.config.snapshot_path = self.form.textedit_snapshot_path.text()
+
+    def get_formatted_comma_separated_string(self, ufList: list) -> str:
```
imho this one is a bit to verbose in the name of the function and not verbose enough/not pep8-complient in the name of the parameter

ohare93 replied:
Renamed and changed to a static function

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,119 @@
+    def set_sort_method(self, set_to):
+        self.config.export_deck_sort_methods = set_to
+    
+    def setup(self):
```
ah, this actually does sorting. How about another new name 😛  - `sort_with([sort_methods])` to make it clear that we actually run the sort here with the given sort method

Stvad commented:
> Jesus that is not how it looks on my screen! 😅 Here's how it looks to me:
> 
> > ![image](https://user-images.githubusercontent.com/1471350/67186199-db139780-f3e7-11e9-9946-a97aaa493d8d.png)

Hmm, some of it is platform differences probably, but the proportions differences are weird 🤔

ohare93 commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,76 @@
+            super().accept()
+
+    def reject(self):
+        super().reject()
```
Removed

Stvad commented on test/anki/hook_vendor_spec.py:
```diff
@@ -8,13 +8,15 @@
+
 def setup_vendor(automated_snapshot):
-    hook_manager = MagicMock()
+    h = MagicMock()
```
why these renamings?

ohare93 replied:
PEP violation, they were "shadow names" of the function which called them. Open to a better name for those object, but it's just tests so I didn't care much.

Stvad replied:
good names are important everywhere ;) 

how about `hook_vendor_mock` `hook_manager_mock`?

Stvad commented:
> I have been quiet lately as I've been busy, but can get back to this soon.

👍 

> I also switched over to PyCharm, which has helped in development but been a time sink. 

PyCharm is nice :). Why was it a time sink? 

> I've checked in fixed for the Python violations. Any remaining were in your original code, or are necessary ones to be ignored (accessing a private module). Let me know if you disagree with that assessment 😉

Feel free to fix any violations in my code that you see. I believe i haven't added any of them consciously besides the following 2 (should probably add this to Contributing.): 
* Anki interaction points require violations at times, as anki codebase does not follow pep8
* My preferred line lenght is ~120 and not 80.

> Been over all your comments, have a few more tasks to get done and then let's see how pullable this monster is. I admit, I am starting to get fed up with it 😅 😏

We're getting there ;) I appreciate you working on the feedback and contributing in general!

Stvad commented:
Btw if there are any things that were non-obvious/you think can get other people with contributions - please add them to Contributing.md

ohare93 commented:
> PyCharm is nice :). Why was it a time sink?

Just to get used to it, and get my setup working again like I had on VS Code, with auto completion and linting for Anki's code. I still don't have it so I can run/debug directly from PyCharm (I can't work out the damn run config settings needed!) like I had in VS Code, but I can get around that for now.

> Feel free to fix any violations in my code that you see. I believe i haven't added any of them consciously besides the following 2 (should probably add this to Contributing doc):
> 
> * Anki interaction points require violations at times, as anki codebase does not follow pep8
> * My preferred line lenght is ~120 and not 80.

:+1: 

> We're getting there ;) I appreciate you working on the feedback and contributing in general!

:100: :+1: :grin:

ohare93 commented:
Bunch more changes, replied to a lot of comments too. Getting through it! :sweat_smile:

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,62 @@
+
+    def save_values_to_anki(self):
+        for prop in self.Properties:
+            self._config.update({prop.value.config_name: getattr(self, prop.value.config_name)})
```
why not just use the dict set value syntax, like:

`self._config.update[prop.value.config_name] = getattr(self, prop.value.config_name)`

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,39 @@
+    def __init__(self, config: ConfigSettings):
+        self.sort_methods = [
+            ConfigSettings.DeckExportSortMethods(method)
+            for method in config.export_deck_sort_methods
```
No not lambda values. The enum ones. Right now you do the transformation from string in the config to an enum values here, but this should be the responsibility of the config class

Stvad replied:
so here you would just do `self.sort_methods = config.sort_methods` then

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,41 @@
+        ]
+        self.is_reversed = getattr(config, ConfigSettings.Properties.EXPORT_NOTES_REVERSE_ORDER.value.config_name)
+
+    def should_skip_sorting(self):
```
given that it's only used with `not` should probably be inverted and be `should_sort` instead

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,41 @@
+    def sort_notes(self, notes):
+        if not self.should_skip_sorting():
+            notes = sorted(notes, key=self.get_sort_key, reverse=self.is_reversed)
+        elif self.is_reversed:
```
hmm, I believe we've discussed this one in one of previous iterations, I believe this would be clearer if you do sort first and then reverse separately.

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,62 @@
+    NOTE_MODEL_ID = "notemodelid"
+
+
+class ConfigSettings:
```
Adding comment to track tests for this one, as I'm getting lost among all outdated ones 😅

ohare93 replied:
Testing added :wink:

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,102 @@
+        self.sorted_notes = self.note_sorter.sort_notes(self.notes)
+
+
+with describe(NoteSorterTester) as self:
```
Adding this one to track a test for >1 sort method

ohare93 replied:
Added one :wink: as mentioned in comment below

ohare93 commented:
Had a spare half hour, managed to get an example of multi note sorting methods testing. I'm sure there's some Python magic I can do to test every one of the combinations, but my mind is blown for today :hear_no_evil: 

Progress tho!

Stvad commented:
Awesome! :) I don't think all combinations are required ;)

ohare93 commented:
Checking some tests for Config_Settings :+1: 

One test is commented out, as I can not get it to work :thinking: It is to test the saving, however the _config file gets set to a MagicMock from Anki's getConfig function, and I cannot make these damn mocks return the actual value of what I give them :sweat_smile:

Stvad commented:
`_config` is not a file it's a dictionary. So you can set it to an empty dict or something before calling save.

Another option which I'd probably find more preferable is do what we discussed before when there was a guard flag on the `__init__` of `ConfigSettings` - to pass in the dictionary and load things from anki only if it's not provided. In that way you'd be able to do other testing in more realistic fashion as well as your code flow would be same for tests and irl.

So the init method would look like:
```python
    def __init__(self, config_values = None):
        self._config = config_values or mw.addonManager.getConfig(__name__)
        self.load_values()
```

Stvad commented:
Also that would allow to properly engage/test the whole default loading mechanism

ohare93 commented:
> `_config` is not a file it's a dictionary. So you can set it to an empty dict or something before calling save.
> 
> Another option which I'd probably find more preferable is do what we discussed before when there was a guard flag on the `__init__` of `ConfigSettings` - to pass in the dictionary and load things from anki only if it's not provided. In that way you'd be able to do other testing in more realistic fashion as well as your code flow would be same for tests and irl.
> 
> So the init method would look like:
> 
> ```python
>     def __init__(self, config_values = None):
>         self._config = config_values or mw.addonManager.getConfig(__name__)
>         self.load_values()
> ```

Yea good call. Done :+1: Test works happily now

ohare93 commented:
Singleton-ness fixed :+1: I do not believe it actually caused any issues before, but it's always good to stay ahead of ourselves.

Turns out the issue with the automated snapshot was there already, as hooks are only setup when the options was True *on startup*. I have changed the HookVendor to always setup the hook, but for the ArchiveVendor function to check the config setting as it tries to run (it already had access to config, so it was a minor change). Have tested it manually myself, and snapshots work on switching profile now, regardless of what your setting was initially :grin: 

This also made the hook_vendor_spec useless, as there is nothing to test, so I deleted it. Considered adding a similar test for ArchiveVendor now, but it looked hard, and I'd rather get your opinion on the change before going too far :grin:

Stvad commented on crowd_anki/anki/hook_vendor.py:
```diff
@@ -21,9 +21,6 @@ def setup_exporter_hook(self):
     def setup_snapshot_hooks(self):
-        if not self.config.automated_snapshot:
-            return
-
```
why this change?

Stvad replied:
ah, read your other comments

Stvad commented on generate_ui.sh:
```diff
@@ -0,0 +1,3 @@
+echo "Generating UI Files"
```
you should add (shebang)[https://en.wikipedia.org/wiki/Shebang_(Unix)] 

also - you've added the `crowd_anki/user_files/Test/zzzNoteTypeExamples` and other test stuff again - please remove them and consider adding them to your local gitignore

ohare93 replied:
Shebang added :+1: 

I have added test things again? :upside_down_face: I was so careful :sob: but how come I cannot find them? :thinking: Both this changeset list lacks them (in my view), and my branch directory is empty: https://github.com/ll-in-anki/CrowdAnki/tree/sortnotesconfigoption/crowd_anki/user_files :thinking:

Stvad replied:
mm, not sure 😛 they don't seem to be there anymore 😅

Stvad commented on package_plugin.sh:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-pyuic5 ui_files/config.ui -o crowd_anki/config/config_ui.py
+bash generate_ui.sh
```
if you make it executable and add shebang, you should be able to call it similar to the fetch dependencies one

ohare93 replied:
Oh you can do that anyways, I just always use bash for some reason. Changed it to be simpler, done :+1:

Stvad commented on test/anki/hook_vendor_spec.py:
```diff
@@ -1,35 +0,0 @@
-from mamba import description, it
```
why is this deleted?

Stvad replied:
read other comments

Stvad commented on test/config/config_settings_spec.py:
```diff
@@ -35,31 +36,31 @@
+            assert config._config["automated_snapshot"] == config.automated_snapshot
+            assert config._config["snapshot_root_decks"] == config.snapshot_root_decks
+            assert config._config["export_notes_reverse_order"] == config.export_notes_reverse_order
+            assert config._config["export_note_sort_methods"] == config.export_note_sort_methods
```
I think a better (and also easier to write) test would be to verify that this dict gets passed in to the anki mock (as then you can reuse your dict definition above and the code would go the full circle from init to save)

Stvad replied:
Though I remember you've had trouble interacting with that mock, so if you can't figure it out this should be ok as well

Stvad replied:
Ah, on looking at it again now - I think it'd be a good idea to pass in the main window (or even just addonManager probably) as a parameter to constructor, with None as default value (similar to what we do for dict now).

That way you'd be able to pass in the mock you want from here, and do the testing properly

ohare93 replied:
Actually I just realised from your comment, that this does not test the saving. I am checking the ConfigSettings' _config dictionary, that those values are set. Instead I should have been checking if the attributes of ConfigSettings get set based on said dictionary. Changed to your suggestion, and testing properly now:
```

        with it("tries to save"):
            settings = {
                "snapshot_path": "testing",
                "automated_snapshot": True,
                "snapshot_root_decks": ["TestDeck1", "Other"],
                "export_notes_reverse_order": True,
                "export_note_sort_methods": ["notemodel", "guid"]
            }

            config = ConfigSettings(settings)

            config.save()

            assert config._config == settings

            for key in settings:
                assert settings[key] == getattr(config, key)

```

All working :+1: thanks for the comment

Stvad replied:
still may want to verify that save on the actual anki is called.

Stvad replied:
doing getattr actually tests loading and not saving. in save you're copying things over to that map (so it's helpful to assert that it looks like what you'd expect) and then call anki to save the map to file

ohare93 replied:
Ahh I see, I got myself muddled up. This test only worked because the load_values function is called in the constructor of ConfigSettings. Changed to two separate tests, which test both the loading and saving :+1: 

```
        with it("tries to load to initial values in the constructor"):
            settings = {
                "snapshot_path": "testing",
                "automated_snapshot": True,
                "snapshot_root_decks": ["TestDeck1", "Other"],
                "export_notes_reverse_order": True,
                "export_note_sort_methods": ["notemodel", "guid"]
            }

            config = ConfigSettings(init_values=settings)

            assert config._config == settings

            for key in settings:
                assert settings[key] == getattr(config, key)

        with it("tries to save new values to config dictionary"):
            addon_manager_mock = MagicMock()

            old_settings = {
                "snapshot_path": "",
                "automated_snapshot": False,
                "snapshot_root_decks": [],
                "export_notes_reverse_order": False,
                "export_note_sort_methods": []
            }

            new_settings = {
                "snapshot_path": "testing",
                "automated_snapshot": True,
                "snapshot_root_decks": ["TestDeck1", "Other"],
                "export_notes_reverse_order": True,
                "export_note_sort_methods": ["notemodel", "guid"]
            }

            config = ConfigSettings(addon_manager=addon_manager_mock, init_values=old_settings)

            for key in new_settings:
                setattr(config, key, new_settings[key])

            config.save()

            assert config._config == new_settings
            addon_manager_mock.writeConfig.assert_called_once()

            for key in new_settings:
                assert new_settings[key] == config._config[key]

```

Stvad commented:
> Turns out the issue with the automated snapshot was there already, as hooks are only setup when the options was True _on startup_.

That's fair 😛. I picked that one randomly though, as an example of general principle (not a very lucky one apparently 😃). If my model of the world is correct it'd be still broken without the change to use singleton instance everywhere

Stvad commented on crowd_anki/export/anki_exporter_wrapper.py:
```diff
@@ -27,7 +28,7 @@ def __init__(self, collection,
         self.did = deck_id
         self.count = 0  # Todo?
         self.collection = collection
-        self.anki_json_exporter = json_exporter or AnkiJsonExporter(collection)
```
alert - new instance is created. It's worrying that it's easy to make this error 🤔

ohare93 replied:
God damn it. Fixed :sweat_smile:

Stvad commented:
Btw as we're getting close - you should rebase on mainline, as github tells me that there are conflicts ;)

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,76 @@
+            self.config.save()
+            super().accept()
+
+    def ui_initial_setup(self):
```
this one vaguely bothers me seems like a lot of boilerplate, wonder if there is a better way (I don't know one straight away, but...)

Stvad commented:
Giving it another full review now

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,88 @@
+    FLAG = "flag"
+    TAG = "tag"
+    NOTE_MODEL = "notemodel"
+    NOTE_MODEL_ID = "notemodelid"
```
I believe the value should probably match the enum name here (add underscores)

Stvad replied:
um, I'd actually prefer to type and especially to read the version with underscores

ohare93 replied:
This_too!

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,88 @@
+class NoteSortingMethods(Enum):
+    FIELD1 = "field1"
+    FIELD2 = "field2"
+    NO_SORTING = "none"
```
I think the name value should match here to, not sure which way though 😅

ohare93 replied:
If you feel it's alright for the user to type "no_sorting" then simply say so and I shall change it! :100: :+1:  Especially with the helpful text at the side, and a warning when an incorrect value is entered, it should be no issue. I simply dislike such entry systems myself, and thus do not wish to make one like that :smile: but we cannot all have what we want. Final call is yours :+1:

Stvad replied:
eh, I don't have strong arguments either way on this one, the `notemodelname` and such should definitely haven underscores in it thouhg ;)

ohare93 replied:
Alright, I'll leave `none` as it is, and change the others to `note_model_name` and `note_model_id` :+1:

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,88 @@
+        for prop in self.Properties:
+            self._config[prop.value.config_name] = getattr(self, prop.value.config_name)
+
+        mw.addonManager.writeConfig(__name__, self._config)
```
as discussed elsewhere seems like a good idea to pass in the addonManager as parameter to constructor

Stvad replied:
I think it's still ok to import `mw` here as it was before, just don't use it directly in the body of the class, but only in the constructor to initialize addonManager if it's not passed in

so something like:

```python
    def __init__(self, addon_manager=None, init_values=None):
        self.addon_manager = addon_manager or mw.addonManager
```

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,88 @@
+
+        return incorrect_sort_methods
+
+    def handle_empty_textboxes(self):
```
how about `try_infer_values` or something like that?

ohare93 replied:
Sure, done.

Stvad commented on crowd_anki/export/note_sorter.py:
```diff
@@ -0,0 +1,39 @@
+
+class NoteSorter:
+    sorting_definitions = {
+        NoteSortingMethods.NO_SORTING: lambda i: i,
```
is this actually valid? i.e. if we return just the object itself, the sort method is still going to sort it somehow right?

ohare93 replied:
I believe so, but it does not matter because it is not used. That's what should_sort checks for, and skips the sort process if NO_SORTING is selected. If someone sets the selected sort methods to "guid, none" then this sort method will be used, but since there is already a first it will not matter. It'll already be different than your default before this PR, by their own selection, so there should be no issue. Right? :thinking:

Stvad replied:
well tbh, having none in combination with other things does not really make sense to me. But given that we can have it, the only coherent way to implement it is to not change position of items at all if the "none" is provided. I think, given that python sort implementation is stable, this can be achieved by returning some constant value from the key.

Stvad replied:
👍 I'd go with the string that'd explain the reasoning above though, or at least adding a comment

ohare93 replied:
Alright, how'd you like the look of this? :grin: 

```python
class NoteSorter:
    sorting_definitions = {
        # NO_SORTING is a special case which should be ignored by should_sort
        # However it's been kept here for edge cases, and for ease of testing
        NoteSortingMethods.NO_SORTING: lambda i: 1,
        
        NoteSortingMethods.GUID: lambda i: i.anki_object.guid,
        NoteSortingMethods.FLAG: lambda i: i.anki_object.flags,
        NoteSortingMethods.TAG: lambda i: i.anki_object.tags,
        NoteSortingMethods.NOTE_MODEL: lambda i: i.anki_object._model["name"],
        NoteSortingMethods.NOTE_MODEL_ID: lambda i: i.anki_object._model["crowdanki_uuid"],
        NoteSortingMethods.FIELD1: lambda i: i.anki_object.fields[0],
        NoteSortingMethods.FIELD2: lambda i: i.anki_object.fields[1]
    }
```

Stvad replied:
I believe the justification for why having `1` there make sense should be there (the whole thing about stable sorting)

ohare93 replied:
```python
class NoteSorter:
    sorting_definitions = {
        # NO_SORTING is a special case which should be ignored by should_sort
        # However it's been kept here for edge cases, and for ease of testing
        # It returns 1 to ensure stable sorting, so the results will be in their previous order
        NoteSortingMethods.NO_SORTING: lambda i: 1,

        NoteSortingMethods.GUID: lambda i: i.anki_object.guid,
        NoteSortingMethods.FLAG: lambda i: i.anki_object.flags,
        NoteSortingMethods.TAG: lambda i: i.anki_object.tags,
        NoteSortingMethods.NOTE_MODEL_NAME: lambda i: i.anki_object._model["name"],
        NoteSortingMethods.NOTE_MODEL_ID: lambda i: i.anki_object._model["crowdanki_uuid"],
        NoteSortingMethods.FIELD1: lambda i: i.anki_object.fields[0],
        NoteSortingMethods.FIELD2: lambda i: i.anki_object.fields[1]
    }
    
```

:thinking: ?

Stvad commented on test/config/config_settings_spec.py:
```diff
@@ -0,0 +1,65 @@
+            invalid_examples = ["testing", "giud", "dan", "flags", "notemodels", "feild", "card", "note"]
+
+            config.export_note_sort_methods = invalid_examples + valid_sorting_methods
+            assert (len(valid_sorting_methods) + len(invalid_examples)) == len(config.export_note_sort_methods)
```
mm, this seems to just assert that `+` works?

Stvad commented on test/config/config_settings_spec.py:
```diff
@@ -0,0 +1,65 @@
+            config = ConfigSettings()
+
+            valid_sorting_methods = list(NoteSortingMethods.values())
+            assert len(valid_sorting_methods) > 0
```
not sure what is this verifying

Stvad replied:
should be a separate test then, that is clearly focused on that ;)

Stvad commented on test/config/config_settings_spec.py:
```diff
@@ -0,0 +1,65 @@
+
+            assert results == invalid_examples
+
+        with it("should set the empty textboxes to their default values"):
```
they are not textboxes 😛

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,148 @@
+    (NoteSortingMethods.FIELD2, test_fields)
+]
+
+test_multikey_notemodel_guid = [(notemodel, guid) for notemodel in test_notemodels for guid in test_guids]
```
this seems to create `N^2` number of pairs, did you want something like `zip` instead?

Stvad commented on crowd_anki/config/config_settings.py:
```diff
@@ -0,0 +1,88 @@
+    GUID = "guid"
+    FLAG = "flag"
+    TAG = "tag"
+    NOTE_MODEL = "notemodel"
```
also looking at the actual sort method - this should be NOTE_MODEL_NAME

Stvad replied:
underscores!

ohare93 replied:
Consider_it_done!

Stvad replied:
🎉

Stvad commented on test/export/note_sorter_spec.py:
```diff
@@ -0,0 +1,148 @@
+        return note
+
+    @staticmethod
+    def get_multikey_note_mock(i):
```
I'm confused on why multi-key setup is different from single key? you still want the same values set, it's just the order in which you expect them sorted would presumably change

Stvad replied:
I'm not sure if the N^2 setup is required in a first place. You're really testing if the sort method is called and solves your scenario in principle and not if the sort method is correct in itself. So I think testing all possible combinations is redundant for that purpose.

ohare93 replied:
Sure, I agree. Though I'd still need a different result at the end to compare it to, and thus would still need a different function to set them to. I agree this part of the class could use some type of tidying up, but there's nothing obvious that jumps out just now :thinking: I suggest we come back to this and leave it like this for now

ohare93 commented:
> Giving it another full review now

Sounds great :+1: I shall have time to review your review tomorrow or so, so take your time.

> Btw as we're getting close - you should rebase on mainline, as github tells me that there are conflicts ;)

Merged in the changes from master :+1: Was just the Pipfile.lock that was the issue, so no big deal.

Stvad commented on test/config/config_settings_spec.py:
```diff
@@ -56,11 +59,41 @@
+            addon_manager_mock.writeConfig.assert_called_once()
+
+            for key in new_settings:
+                assert new_settings[key] == config._config[key]
```
this seems to duplicate the assertion above (of just comparing dicts)

ohare93 replied:
Looking online says you're right, that a straight up comparison `assert config._config == new_settings` is fine :+1: I am used to not trusting such comparisons, and doing extra work myself :sweat_smile: Deleted

Stvad commented:
> Sooo close! 😁

Indeed, I feel like we can conclude things today :)

ohare93 commented:
> > Sooo close! grin
> 
> Indeed, I feel like we can conclude things today :)

:100: :ok_hand: 

Minor changes complete. I am available for the next 3 hours, then I am gone for the day :slightly_smiling_face: If there are any small change left let me know and I shall get it done! :+1:

Stvad commented:
Going to sleep now myself. Gonna take a final look tomorrow, and hopefully merge it :)

Stvad commented:
Ah, I believe you haven't actually addressed the UI improvements we've discussed, oh well, I'll create an issue to track that.

Stvad commented on crowd_anki/config/config_dialog.py:
```diff
@@ -0,0 +1,77 @@
+        invalid_values = self.config.find_invalid_config_values()
+        if invalid_values:
+            error_message_popup = QMessageBox(self)
+            message_text = "\" , \"".join(invalid_values)
```
nit: you can combine single and double quotes to make this cleaner (i.e. double quotes can be placed in single one or other way around without escaping)

Stvad commented:
Merged! Thank you for contributing @ohare93 !
